models/uniswap/ethereum/uniswap_v2_ethereum_trades.sql

```diff
@@ -5,7 +5,7 @@
     materialized = 'incremental',
     file_format = 'delta',
     incremental_strategy = 'merge',
-    unique_key = ['block_date', 'blockchain', 'project', 'version', 'tx_hash', 'evt_index'],
+    unique_key = ['block_date', 'blockchain', 'project', 'version', 'tx_hash', 'evt_index', 'token_bought_address', 'token_sold_address'],
```

Comment:
adding two new unique keys is concerning, as we use the exact same set of unique keys in all projects that feed into `dex.trades` and it has worked well universally. why are we adding two more here?

Reply:
This is needed because this new implementation potentially decodes the swap event into two records - token0 for token1 and token1 for token0. See failed checks right before this commit for more context.

An alternative approach would be to aggregate those two records into a single one, but it would make the code quite a bit more complex/verbose (we'd have to figure out what is the token sold/bought based on the net amounts in/out of tokens 0/1) and potentially cause issues with negative values (still need to confirm this), so that's the tradeoff

Reply:
It does seem that the alternative approach I mentioned would result in negative values. Here's an example of an unusual [swap event](https://etherscan.io/tx/0x2bc47e6c0781485058328e32b6edb37520bb23a50d9b968bde4b04471e50436b#eventlog) (index 128) : 

```
amount0In : 659774361155999
amount1In : 250390
amount0Out : 100000355
amount1Out : 0
```

The net amounts are such that the pool's balance of **both** tokens increases. An heuristic that tried to infer the tokens sold/bought would result in both tokens as `token_sold` (ie sold by the trader to the pool). Still unclear to me why this happens, but I assume it's something along the lines of someone having transferred token0 to the pool before the actual trade occurred, and the Swap event just catching up with the state.

Reply:
I've added the two new keys to the dex.trades schema in 1b9c1fb